### PR TITLE
Specialize eps(::IEEEFloat)

### PR DESCRIPTION
**Pull request by mikmoore**
`eps(::IEEEFloat)` can be made notably faster. I doubt this function appears in performance-critical places, but better is better and this new version is branch-free.

```julia-repl
julia> using BenchmarkTools

julia> x = randn(2^10);

julia> @btime map(eps,$x); # v1.9.1
  2.067 μs (1 allocation: 8.12 KiB)

julia> @btime map(eps,$x); # this PR
  613.095 ns (1 allocation: 8.12 KiB)
```

A couple probably-negligible caveats that arise from the use of floating point arithmetic in this implementation:
- The previous version always returned Julia's canonical `NaN` when the result would be NaN. In this PR, the NaN payload depends on the payload behavior of the hardware and the input.
- The result is vulnerable to flush-to-zero floating point environments, in which case subnormal results will be flushed to `zero(T)`. The usability of a subnormal result amidst flushing mode is questionable, as is the well-posed-ness of `eps`'s semantics in such a situation. *Note that the existing implementation has this same vulnerability (due to `ldexp`), though only when the input to `eps` is normal and the result would be subnormal (only with an input of zero or a subnormal is a correct subnormal output produced, assuming the flushing occurs during arithmetic but not storage).*

Canonicalizing `NaN` would require a simple `ifelse` switch on the result, though I can't imagine we care. Guaranteeing a proper subnormal result in flush-to-zero mode may be more challenging (and may warrant a branch) and is only handled slightly more correctly by the preexisting implementation. I imagine that flushing breaks a lot more functions than just this one, anyway.

### CODE REVIEW

File: base/float.jl [hunk 1]
```diff
@@ -1014,13 +1014,24 @@ isodd(x::AbstractFloat) = isinteger(x) && abs(x) ≤ maxintfloat(x) && isodd(Int
...
     floatmax(::Type{Float64}) = $(bitcast(Float64, 0x7fefffffffffffff))
 
-    eps(x::AbstractFloat) = isfinite(x) ? abs(x) >= floatmin(x) ? ldexp(eps(typeof(x)), exponent(x)) : nextfloat(zero(x)) : oftype(x, NaN)
     eps(::Type{Float16}) = $(bitcast(Float16, 0x1400))
     eps(::Type{Float32}) = $(bitcast(Float32, 0x34000000))
     eps(::Type{Float64}) = $(bitcast(Float64, 0x3cb0000000000000))
     eps() = eps(Float64)
 end
 
+eps(x::AbstractFloat) = isfinite(x) ? abs(x) >= floatmin(x) ? ldexp(eps(typeof(x)), exponent(x)) : nextfloat(zero(x)) : oftype(x, NaN)
```
Comment (oscardssmith) [inline base/float.jl:1023]: I don't think moving this out of the `@eval` does anything (but on the other hand, I don't think that `@eval` does anything either)..

**Review by oscardssmith — APPROVED** [review]

File: base/float.jl (hunk 1 above)
Comment (mikmoore) [reply base/float.jl:1023]: Agreed -- I was just keeping them together. There are a few things that don't take advantage of that `@eval`, and the things that do would presumably be resolved by constant propagation anyway.

**Comment by oscardssmith** [comment]
llvm does not preserve which NaN gets produced, so imo without the normalization is better

**Comment by mikmoore** [comment]
It looks like there are no outstanding objections. Bumping this for merge consideration. I would try to re-run the CI but I don't know how, though the failures didn't look relevant. Someone can re-run those if desired.

**Pull request merged**